runtime/commonMain/src/kotlinx/serialization/Annotations.kt

```diff
@@ -10,60 +10,156 @@ import kotlinx.serialization.json.*
+ * Box.serializer(Box.serializer(Int.serializer())) // Returns serializer for Box<Box<Int>>
+ * ```
+ *
+ * ### Implementation details
+ *
+ * In order to generate `serializer` method that is not a method on the particular instance, the class should have a companion object, either named or unnamed.
+ * Companion object is generated by the plugin if it is not declared, effectively exposing both companion and `serializer()` method to class ABI.
+ * If companion object already exists, only `serializer` method will be generated.
+ * If `serializer` method already exists, it will be matched against the desired signature `serializer(vararg type params): KSerializer<T>`.
+ * On successful match, the existing method will be used as [KSerializer] source instead of auto-generated one.
+ * On unsuccessful match, a compiler error occurs.
  *
  * @see UseSerializers
+ * @see Serializer
  */
 @Target(AnnotationTarget.PROPERTY, AnnotationTarget.CLASS, AnnotationTarget.TYPE)
+@Retention(AnnotationRetention.BINARY)
 public annotation class Serializable(
-    val with: KClass<out KSerializer<*>> = KSerializer::class // it means -- use default serializer by default
+    val with: KClass<out KSerializer<*>> = KSerializer::class // Default value indicates that auto-generated serializer is used
 )
 
 /**
- * Instructs plugin to turn this class into serializer for specified class [forClass].
+ * Instructs the serialization plugin to turn this class into serializer for specified class [forClass].
  * However, it would not be used automatically. To apply it on particular class or property,
  * use [Serializable] or [UseSerializers], or [ContextualSerialization] with runtime registration.
  */
 @Target(AnnotationTarget.CLASS)
+@Retention(AnnotationRetention.BINARY)
 public annotation class Serializer(
-    val forClass: KClass<*> // what class to create serializer for
+    val forClass: KClass<*> //  target class to create serializer for
 )
 
 /**
- * Overrides name visible to the runtime part of serialization framework
+ * Overrides the name of the current element in the [SerialDescriptor].
+ *
+ * By default, [SerialDescriptor.serialName] and [SerialDescriptor.getElementName]
+ * are matched with fully-qualified name of the target class and the name of the property respectively.
+ * Applying this annotation changes the visible name to the given [value]:
+ *
+ * ```
+ * package foo
+ *
+ * @Serializable
+ * class Foo(val myInt: Int)
+ *
+ * @Serializable
+ * @SerialName("Bar")
+ * class Bar(@SerialName("int") val myInt: Int)
+ *
+ * fun printNames(descriptor: SerialDescriptor) {
+ *     println(descriptor.serialName)
+ *     println(descriptor.getElementName(0))
+ * }
+ *
+ * // Prints "foo.Foo" and "myInt"
+ * printNames(Foo.serializer().descriptor)
+ *
+ * // Prints "Bar" and "int"
+ * printNames(Bar.serializer().descriptor)
+ * ```
+ * Element names and serial names are used by text-based serial formats in order to encode the name of the class or
```

Comment:
Better to put this in the beginning -- this is what most of the users want to know, introspecting descriptors is an advanced usage pattern.

Also, you may add code with Json to example.